designs/autoscaler.md

```diff
@@ -0,0 +1,526 @@
+
+### Goals and Objectives
+
+As an EKS Anywhere user, I want to:
+
+* create clusters in a streamlined fashion with autoscaling capabilities
+* add autoscaling capabilities to existing EKS-A clusters
+* have clusters automatically scale worker nodes up and down within the set boundaries
+
+### Statement of Scope
+
+#### In Scope
+
+* Include Cluster Autoscaler as an [EKS-A curated package](https://anywhere.eks.amazonaws.com/docs/concepts/packages/).
+* Integrate Cluster Autoscaler to all CAPI providers since the feature is provider agnostic.
+* Integrate autoscaling configuration in cluster spec to make EKS-A cluster autoscaling-aware.
+* Generate an EKS-A opinionated Cluster Autoscaler curated package template with default configurations.
+* Install Cluster Autoscaler controller as a curated package during or after management/workload cluster creation.
+* Support upstream k8s Autoscaler installed/managed outside of EKS-A Curated Package.
+* Update the autoscaling configuration to set up the min/max number of nodes in each worker node group.
+
+#### Out of Scope
+
+* Compare cluster autoscaler products (e.g. [Karpenter](https://karpenter.sh/))
+* Autoscale at the pod level
+* Control Plane nodes autoscaling
+
+#### Future Scope
+
+* Support other autoscaling solutions (e.g. [Karpenter](https://karpenter.sh/)).
+
+## Overview of Solution
+
+We provide a solution to add autoscaling configuration in the EKS Anywhere API, which marks the worker node groups to be Cluster Autoscaler compatible. EKS Anywhere user can then install Cluster Autoscaler in any EKS-A cluster with proper configuration, and expect node autoscaling to work for the associated resource group.
+
+1. Introduce an optional type `AutoScalingConfiguration` in the cluster and populate it into each worker node group configuration.
+2. If `AutoScalingConfiguration` is specified (not nil) for a worker node group, EKS-A converts its fields to CAPI autoscaler annotations and applies them to the underlying CAPI Machine Deployment resource. In this way, the specified worker node group will be discoverable by the Cluster Autoscaler controller once installed.
+3. Install Cluster Autoscaler either (A) as an EKS-A curated package, or (B) from upstream; with configuration that uses Cluster API as cloud provider, and auto-discovers node groups in the target cluster.
+
+### API Design
+
+```go
+type WorkerNodeGroupConfiguration struct {
+    // Count defines the number of desired worker nodes.
+    Count *int `json:"count,omitempty"`
+    
+    // AutoScalingConfiguration defines the auto scaling configuration
+    // +optional
+    AutoScalingConfiguration *AutoScalingConfiguration `json:"autoscalingConfiguration,omitempty"`
+    ...
+}
+
+type AutoScalingConfiguration struct {
+    // MinCount defines the minimum number of nodes for the associated resource group.
+    // +optional
+    MinCount int `json:"minCount,omitempty"`
```

Comment:
Just curious, but on line 98, it seem that the cluster autoscaler already uses the terms "min-size" and "max-size". Would it be worthwhile to maintain consistency and use minSize / maxSize in our code?

Reply:
eks-a spec is the only one we expose to the user, not CAPI. It is more important to stay in consistent w eks-a field rather than underlying CAPI field. Similar to `Count`, underlying it's named `replicas` in CAPI.